### PR TITLE
add basic interfaces and structures for match cycles

### PR DESCRIPTION
**Pull request by Felk**

### CODE REVIEW

File: Match/FeatureSet.cs [hunk 1]
```diff
@@ -0,0 +1,37 @@
+using System.Collections.Immutable;
+
+namespace Match
+{
+    public enum Generation
+    {
+        Gen1, Gen2, Gen3, Gen4, Gen5, Gen6, Gen7, Gen8
```
Comment (red031000) [inline Match/FeatureSet.cs:7]: start enum at 1

File: Match/FeatureSet.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
+using System.Collections.Immutable;
+
+namespace Match
+{
+    public enum Generation
```
Comment (red031000) [inline Match/FeatureSet.cs:5]: put enums in a class so there are no naming conflicts?

File: Match/IMatchCycle.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+    {
+        /// information on what features the match implementation supports
+        public FeatureSet FeatureSet { get; }
+
+        /// Sets up a match for the provided match info,
+        /// and returns a delegate to perform the prepared match.
+        public Task<Perform> SetUp(MatchInfo matchInfo, CancellationToken? token = null);
+
+        /// A delegate that performs a match that was prepared with <see cref="IMatchCycle.SetUp"/>.
+        delegate Task<MatchResult> Perform(CancellationToken? token = null);
```
Comment (red031000) [inline Match/IMatchCycle.cs:16]: delegate vs event?

File: Match/MatchInfo.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+    public record MatchInfo(
+        IImmutableList<Pokemon> TeamBlue,
+        IImmutableList<Pokemon> TeamRed)
+    {
+        public string? Stage { get; init; }
+    // dotnet-format is drunk, see: https://github.com/dotnet/format/issues/805
+    public string? FieldEffect { get; init; }
+    public float Speed { get; init; } = 1f;
+}
```
Comment (red031000) [inline Match/MatchInfo.cs:18]: bracket alignment?

**Review by red031000 — CHANGES_REQUESTED** [review]
match tests?

File: Match/MatchInfo.cs (hunk 4 above)
Comment (Felk) [reply Match/MatchInfo.cs:18]: see comment above, dotnet is doing weird stuff. 
If that's preferable I can fix it by marking this file as generated code in the editorconfig file instead, so dotnet-format won't touch it.

File: Match/FeatureSet.cs (hunk 2 above)
Comment (Felk) [reply Match/FeatureSet.cs:5]: Is there something also called `Generation` I should be aware of?

File: Match/IMatchCycle.cs (hunk 3 above)
Comment (Felk) [reply Match/IMatchCycle.cs:16]: not sure what you mean

Comment (Felk) [reply Match/IMatchCycle.cs:16]: The reason this isn't just a second interface method is to make sure nobody could call Perform before SetUp

File: Match/FeatureSet.cs (hunk 2 above)
Comment (red031000) [reply Match/FeatureSet.cs:5]: probably should have put this on features instead, incase we want to use it to enable run stuff or smth like that

File: Match/IMatchCycle.cs (hunk 3 above)
Comment (red031000) [reply Match/IMatchCycle.cs:16]: I mean is an event preferable to a delegate here

File: Match/MatchInfo.cs (hunk 4 above)
Comment (red031000) [reply Match/MatchInfo.cs:18]: no that should be fine

File: Match/IMatchCycle.cs (hunk 3 above)
Comment (Felk) [reply Match/IMatchCycle.cs:16]: after a lengthly discord discussion we agreed that this is confusing and should just be two interface methods

**Review by red031000 — APPROVED** [review]

**Pull request merged**